packages/veritone-widgets/src/redux/modules/mediaDetails/index.js

```diff
@@ -0,0 +1,657 @@
+    return {
+      ...state,
+      [widgetId]: {
+        ...state[widgetId],
+        success: !(warn || error) || null,
+        error: error ? errorMessage : null,
+        warning: warn || null,
+        engineCategories: payload
+      }
+    };
+  },
+  [LOAD_ENGINE_RESULTS](state, { payload, meta: { widgetId } }) {
+    let resultRequests =
+      state[widgetId].engineResultRequestsByEngineId[payload.engineId] || [];
+    let requestInsertIndex = findLastIndex(resultRequests, request => {
+      return request.stopOffsetMs < payload.startOffsetMs;
+    });
+    resultRequests = [
+      ...resultRequests.slice(0, requestInsertIndex + 1),
+      {
+        startOffsetMs: payload.startOffsetMs,
+        stopOffsetMs: payload.stopOffsetMs,
+        status: 'FETCHING'
+      },
+      ...resultRequests.slice(requestInsertIndex + 1)
+    ];
+
+    let engineResults =
+      state[widgetId].engineResultsByEngineId[payload.engineId] || [];
+    let resultInsertIndex = findLastIndex(engineResults, result => {
+      return (
+        result.series[result.series.length - 1].startTimeMs <=
+        payload.startOffsetMs
+      );
+    });
+
+    return {
+      ...state,
+      [widgetId]: {
+        ...state[widgetId],
+        engineResultRequestsByEngineId: {
+          ...state[widgetId].engineResultRequestsByEngineId,
+          [payload.engineId]: [...resultRequests]
+        },
+        engineResultsByEngineId: {
+          ...state[widgetId].engineResultsByEngineId,
+          [payload.engineId]: [
+            ...engineResults.slice(0, resultInsertIndex + 1),
+            {
+              startOffsetMs: payload.startOffsetMs,
+              stopOffsetMs: payload.stopOffsetMs,
+              status: 'FETCHING'
+            },
+            ...engineResults.slice(resultInsertIndex + 1)
+          ]
+        }
+      }
+    };
+  },
+  [LOAD_ENGINE_RESULTS_COMPLETE](
+    state,
+    { payload, meta: { warn, error, widgetId, startOffsetMs, stopOffsetMs } }
+  ) {
+    const errorMessage = get(error, 'message', error);
+
+    let previousResultsByEngineId =
+      state[widgetId].engineResultsByEngineId || {};
+    let engineResultRequestsById =
+      state[widgetId].engineResultRequestsByEngineId;
+    // It is possible results were requested by
+    const resultsGroupedByEngineId = groupBy(payload, 'engineId');
+    forEach(resultsGroupedByEngineId, (results, engineId) => {
+      if (!previousResultsByEngineId[engineId]) {
+        // Data hasn't been retrieved for this engineId yet
+        previousResultsByEngineId[engineId] = map(results, 'jsondata');
+      } else {
+        // New results need to be merged with previously fetched results
+        let insertionIndex = findIndex(previousResultsByEngineId[engineId], {
+          startOffsetMs,
+          stopOffsetMs,
+          status: 'FETCHING'
+        });
+
+        // TODO: fitler out any duplicate data that overflows time chunks.
+        previousResultsByEngineId[engineId] = [
```

Comment:
can you abstract this into a function? you're doing the same thing in multiple places.

something like - addEngineResult(data, start, end) {} perhaps?

Reply:
@ChrisPelletier